modules/build/src/main/scala/scala/build/ResourceMapper.scala

```diff
@@ -0,0 +1,122 @@
+package scala.build
+
+import scala.collection.Map
+import scala.jdk.CollectionConverters._
+import scala.build.internal.Constants
+
+trait BaseResourceMapper {
+  private def readMappingIfExists(mappingFile: os.Path): Option[Map[os.Path, os.Path]] =
+    if (os.exists(mappingFile)) {
+      val mappingContent = os.read(mappingFile)
+      if (mappingContent.isEmpty()) Some(Map.empty)
+      else Some(
+        mappingContent.split('\n').map { pair =>
+          val sep = pair.split(',')
+          (os.Path(sep(0)), os.Path(sep(1)))
+        }.toMap
+      )
+    }
+    else None
+
+  private def writeMapping(mappingPath: os.Path, mapping: Map[os.Path, os.Path]) = {
+    val mappingContent = mapping.map { case (inputPath, outputPath) =>
+      s"$inputPath,$outputPath"
+    }.mkString("\n")
+    os.write.over(mappingPath, mappingContent)
+  }
+
+  protected def copyResourcesToDirWithMapping(
+    build: Build.Successful,
+    mappingFilePath: os.Path,
+    mappingFunction: Build.Successful => Map[os.Path, os.Path]
+  ): Unit = {
```

Comment:
Wouldn't the signature be clearer with these arguments? (`mappingFunction` can be applied by the caller, and then there's no need of the whole build here, only the output dir)
```suggestion
  protected def copyResourcesToDirWithMapping(
    output: os.Path,
    mappingFilePath: os.Path,
    newMapping: Map[os.Path, os.Path]
  ): Unit = {
```